Add PCH builds via 'SFML_ENABLE_PCH' CMake option

vittorioromeo opened the pull request:
## Description

Enable PCH via CMake. Achieves roughly a 2x fresh compilation speedup on my machine, using MSYS2/MinGW. Looking for feedback on the direction and approach.

---

<details>
<summary>Machine Specs</summary>

- GIGABYTE Z390 AORUS MASTER-CF
- i9-9900k (stock settings)
- Corsair CMK16GX4M2B3000C15, 4x8GB sticks (XMP enabled)

</details>

---

<details>
<summary>Benchmark</summary>

Methodology:

```
cd build
rm -Rf src test examples

cmake .. -DCMAKE_BUILD_TYPE=Debug -GNinja -DSFML_BUILD_EXAMPLES=1 \ 
    -DCMAKE_EXPORT_COMPILE_COMMANDS=1 -DCMAKE_CXX_COMPILER=clang++ \
    -DCMAKE_CXX_FLAGS="-ftime-trace"

time ninja
```

Results without this PR:
- `ninja  0.00s user 0.01s system 0% cpu 5.599 total`

Results with this PR:
- `ninja  0.00s user 0.01s system 0% cpu 9.586 total`

</details>

---

This PR is related to the issue #1893.

## Tasks

* [ ] Tested on Linux
* [x] Tested on Windows
* [ ] Tested on macOS
* [ ] Tested on iOS
* [ ] Tested on Android

## How to test this PR?

Not sure yet.

vittorioromeo commented:
@binary1248: any idea why CI is unhappy with the changes? The GCC error is particularly weird...

binary1248 commented:
I don't have any experience using PCH, definitely not with CMake.

sam20908 commented:
Did you ensure that the targets reusing the PCH have the same compiler definitions, options, and flags?

Quoted from CMake documentation:
> The REUSE_FROM form requires the same set of compiler options, compiler flags and compiler definitions for both <target> and <other_target>. Some compilers (e.g. GCC) may issue a warning if the precompiled header file cannot be used (-Winvalid-pch).

vittorioromeo commented:
> Did you ensure that the targets reusing the PCH have the same compiler definitions, options, and flags?
> 
> Quoted from CMake documentation:
> 
> > The REUSE_FROM form requires the same set of compiler options, compiler flags and compiler definitions for both  and <other_target>. Some compilers (e.g. GCC) may issue a warning if the precompiled header file cannot be used (-Winvalid-pch).

There are no targets reusing the PCH, I've commented all of that out... The CI still fails with cryptic errors even when enabling PCH just for `sfml-system`.

vittorioromeo commented:
Hah, yes! That was the issue. Removing `-Wpedantic` from `CompilerWarnings.cmake` makes PCH work! I don't understand why though, as we do call `set_file_warnings` on the PCH header. Does anybody want to investigate this further? @ChrisThrasher

vittorioromeo commented:
I solved with 

```cmake
target_precompile_headers(sfml-system PRIVATE "${CMAKE_SOURCE_DIR}/src/SFML/PCH.hpp")
set_file_warnings("${CMAKE_BINARY_DIR}/src/SFML/System/CMakeFiles/sfml-system.dir/cmake_pch.hxx.cxx")
```

but it is really ugly...

ChrisThrasher commented:
Precompiled headers require CMake 3.16. I see this PR doesn't change the minimum CMake version requirement. Is it possible that some jobs are failing due to old CMake versions?

vittorioromeo commented:
At this point in time, PCH files seem to work only for Linux and MinGW, and only when SFML is built as a static library. For Visual Studio, the issue is:

```
CMake Error at cmake/Macros.cmake:[69](https://github.com/SFML/SFML/actions/runs/3165984825/jobs/5155401526#step:8:70) (add_library):
  PRECOMPILE_HEADERS_REUSE_FROM property is set on target ("sfml-window").
  Reusable precompile headers requires the COMPILE_PDB_NAME property to have
  the value "sfml-system"

Call Stack (most recent call first):
  src/SFML/Window/CMakeLists.txt:255 (sfml_add_library)
```

So to make it work, `COMPILE_PDB_NAME` has to be the same as the PCH target we are reusing the PCH from.

Regarding shared library builds: I cannot make PCH reuse work because of differing macro defines for `SFML_XXX_EXPORTS`. All targets must have the same macros. I can try having a per-library PCH, but I'm not sure if it's worth it.

vittorioromeo commented:
I have no idea what the MacOS issue is about:

```
CMake Error in src/SFML/Window/CMakeLists.txt:
  Unable to resolve full path of PCH-header
  '/Users/runner/work/SFML/SFML/build/src/SFML/System/CMakeFiles/sfml-system.dir/cmake_pch.objc.h'
  assigned to target sfml-window, although its path is supposed to be known!
```

vittorioromeo commented:
I have no idea what the MacOS issue is about:

```
CMake Error in src/SFML/Window/CMakeLists.txt:
  Unable to resolve full path of PCH-header
  '/Users/runner/work/SFML/SFML/build/src/SFML/System/CMakeFiles/sfml-system.dir/cmake_pch.objc.h'
  assigned to target sfml-window, although its path is supposed to be known!
```

vittorioromeo commented:
@ChrisThrasher @eXpl0it3r @binary1248 @Bromeon please opine. The speedup on MinGW/Linux is quite large, a fresh compile is at minimum 2x faster on my machine. A few points:

- PCH builds are disabled by default. Can be enabled with the `SFML_ENABLE_PCH` option.

- PCH builds are only supported on Windows/Linux. MacOS is not working for some weird reason and I cannot debug it easily.

- PCH builds are only supported for static builds of SFML. Shared library support is not worth it because the PCH cannot be reused due to mismatching macro definitions of `SFML_XXX_EXPORTS` between targets.

- It would be nice to have "static + PCH" CI jobs for Linux, MinGW, and MSVC.

ChrisThrasher commented on cmake/Macros.cmake:
```diff
@@ -281,11 +300,21 @@ macro(sfml_add_example target)
         add_executable(${target} ${target_input})
     endif()
 
+    # enable C++17 support
+    target_compile_features(${target} PUBLIC cxx_std_17)
```
Why add this? It's redundant since this requirement is already picked up when these targets depend on `SFML::` targets. I removed this a PR back in like February.

vittorioromeo replied:
Will try removing, might be unnecessary.

ChrisThrasher commented on cmake/Macros.cmake:
```diff
@@ -281,11 +300,21 @@ macro(sfml_add_example target)
+    endif()
+
     set_target_warnings(${target})
+    set_public_symbols_hidden(${target})
```
Why do we need to hide symbols in executable targets? I thought that was only meaningful for library.

vittorioromeo replied:
PCH reuse requires the flags to be exactly the same for all projects using the same PCH.

ChrisThrasher commented on cmake/Macros.cmake:
```diff
@@ -322,13 +351,23 @@ function(sfml_add_test target SOURCES DEPENDS)
     add_executable(${target} ${SOURCES})
 
+    # enable C++17 support
+    target_compile_features(${target} PUBLIC cxx_std_17)
```
See above comment about this being redundant.

ChrisThrasher reviewed (changes requested):
In the ideal case, is it possible to support pre-compiled headers on all platforms or does CMake simply not allow that?

What do you think about automatically enabling PCH where possible so that we're not expanding the configuration space of the library? That would let us stop worrying about adding CI jobs and the exponential (polynomial?) expansion of jobs that entails.

ChrisThrasher commented on src/SFML/PCH.hpp:
```diff
@@ -0,0 +1,63 @@
+
+#endif // SFML_SYSTEM_WINDOWS
+
+#include <SFML/Graphics/Rect.hpp>
```
What happens if somebody doesn't build the Graphics module? Will this still work since all headers are accessible no matter whether or not all modules are built?

vittorioromeo replied:
I will circumvent the issue right now by removing this line.

ChrisThrasher commented:
Here are my results testing this PR. I'm using an Ubuntu 22 VM running in macOS with 6 CPUs and 16 GB of memory.

---

`master` (939ec267dfa98e97438a5b76f79317c7e0738a92)
```
cmake -B build -DSFML_BUILD_EXAMPLES=ON -DSFML_BUILD_TEST_SUITE=ON -DBUILD_SHARED_LIBS=OFF
time cmake --build build -j6
```

```
Total Test time (real) =   0.35 sec
[100%] Built target runtests

real	0m13.779s
user	1m1.117s
sys	0m6.092s
```

---

`feature/use-pch` (ec0a03610d22d55fe0ed4916bf0c1dbb9076352e)
```
cmake -B build -DSFML_BUILD_EXAMPLES=ON -DSFML_BUILD_TEST_SUITE=ON -DBUILD_SHARED_LIBS=OFF -DSFML_ENABLE_PCH=ON
time cmake --build build -j6
```

```
Total Test time (real) =   0.35 sec
[100%] Built target runtests

real    0m12.228s
user    0m46.263s
sys    0m7.278s
```

vittorioromeo commented:
My results with MSYS2/MinGW on my Windows 10 x64 laptop:

```
cmake .. -GNinja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS="-ftime-trace" -DSFML_BUILD_TEST_SUITE=1 -DSFML_BUILD_EXAMPLES=1 -DSFML_BUILD_DOC=0 -DBUILD_SHARED_LIBS=0 -DSFML_ENABLE_PCH=0
```

```
time ninja
ninja  0.01s user 0.00s system 0% cpu 31.183 total
```

---

```
cmake .. -GNinja -DCMAKE_C_COMPILER=clang -DCMAKE_CXX_COMPILER=clang++ -DCMAKE_CXX_FLAGS="-ftime-trace" -DSFML_BUILD_TEST_SUITE=1 -DSFML_BUILD_EXAMPLES=1 -DSFML_BUILD_DOC=0 -DBUILD_SHARED_LIBS=0 -DSFML_ENABLE_PCH=1
```

```
time ninja
ninja  0.00s user 0.01s system 0% cpu 22.085 total
```

Bromeon commented on src/SFML/PCH.hpp:
```diff
@@ -0,0 +1,62 @@
+////////////////////////////////////////////////////////////
+//
+// SFML - Simple and Fast Multimedia Library
+// Copyright (C) 2007-2021 Laurent Gomila (laurent@sfml-dev.org)
```
Should be 2022.
Likely also affects the source from which the license header was copy-pasted.

Bromeon commented on src/SFML/System/CMakeLists.txt:
```diff
@@ -67,6 +67,12 @@ endif()
 
+# enable precompiled headers
+if (SFML_ENABLE_PCH)
+    message(VERBOSE "enabling PCH for SFML library 'sfml-system' (reusable)")
```
What is meant with `(reusable)` here? Is there an easy way to clarify the message?

Bromeon commented on cmake/Macros.cmake:
```diff
@@ -80,6 +88,13 @@ macro(sfml_add_library module)
+    set_public_symbols_hidden(${target})
+
+    # enable precompiled headers
+    if (SFML_ENABLE_PCH AND (NOT ${target} STREQUAL "sfml-system"))
+        message(VERBOSE "enabling PCH for SFML library '${target}'")
+        target_precompile_headers(${target} REUSE_FROM sfml-system)
+    endif()
```
Small nitpick: `if(` doesn't have space in other places.

I like how you log the use of PCHs thoroughly, so that users can see what exactly happens!

Bromeon reviewed (commented):
This is a very nice addition to SFML, thanks a lot for this PR! 👍

I'm not sure if we should switch CI over to use PCH=on; probably a lot of SFML users would still use the classic setup, and I think it's important we have the confidence that the basic configuration is not broken.

The combinatoric growth is a problem; there are a few options though:
* simply add more jobs
   * as long as we can still run them in parallel, it shouldn't take much longer (not sure about initial PCH compile times)
* run CI with PCH=on only on rare periodic intervals
   * not a fan of this, as it makes it hard to link it to changes
* only run PCH when CMake is changed
   * C++ code ideally shouldn't break PCH. would however need quite some extra GitHub Actions logic

vittorioromeo commented:
> This is a very nice addition to SFML, thanks a lot for this PR! 👍

Thanks!

> I'm not sure if we should switch CI over to use PCH=on; probably a lot of SFML users would still use the classic setup, and I think it's important we have the confidence that the basic configuration is not broken.

Honestly, this feature is mostly to speed up builds for people working on SFML's codebase itself. Even if it's never consumed by our users, it's still beneficial for us to have as part of our development cycle, and potentially to speed up CI.

My personal preference would be to enable PCH=on on all our CI except one job that tests a non-PCH build just to make sure it still works properly -- one of the reasons to have PCH is to speed up the whole CI process.

Happy to hear what others think.

Bromeon commented:
> My personal preference would be to enable PCH=on on all our CI except one job that tests a non-PCH build just to make sure it still works properly -- one of the reasons to have PCH is to speed up the whole CI process.

That sounds reasonable to me. 👍

vittorioromeo commented:
> I assume miscompilations due to PCH are no longer the dreadful thing they were 1-2 decades ago? 😄

My personal experience with PCH has been very positive in this aspect, I never had any miscompilation issue. Also, as you can see by the pain I went through, any sort of inconsistency between files causes a total compilation failure -- PCH needs every flag to be exactly the same in order to avoid weird issues. I think we should be pretty safe!

vittorioromeo commented:
I would suggest merging this PR as-is, then having a separate PR to enable PCH in CI.

Bromeon commented on src/SFML/PCH.hpp:
```diff
@@ -0,0 +1,62 @@
+
+#ifdef SFML_SYSTEM_WINDOWS
+
+#define UNICODE  1
+#define _UNICODE 1
```
This cannot lead to issues somewhere down the line? Or should they be `#undef`ed after including the Windows headers?

vittorioromeo replied:
This matches the same definitions in the CMakeLists, which are never undefined. It should not cause any issue:

```cmake
if(SFML_OS_WINDOWS)
    set(PLATFORM_SRC
        ${SRCROOT}/Win32/CursorImpl.hpp
        ...
    )
    source_group("windows" FILES ${PLATFORM_SRC})

    # make sure that we use the Unicode version of the Win API functions
    add_definitions(-DUNICODE -D_UNICODE)
```

vittorioromeo replied:
@Bromeon let me know if you're happy with that before merging :)

Bromeon replied:
Ok, fine with me in that case! 🚀

Bromeon reviewed (commented):
One small question, other than that good to go from my side!

ChrisThrasher commented on cmake/Macros.cmake:
```diff
@@ -164,18 +179,22 @@ macro(sfml_add_library module)
+
+                # For PCH builds with PCH reuse, the PDB name must be the same as the target that's being reused
+                set_target_properties(${target} PROPERTIES
+                                    COMPILE_PDB_NAME "sfml-system"
+                                    COMPILE_PDB_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}/lib")
```
```suggestion
                                      COMPILE_PDB_NAME "sfml-system"
                                      COMPILE_PDB_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}/lib")
```

ChrisThrasher commented on cmake/Macros.cmake:
```diff
@@ -164,18 +179,22 @@ macro(sfml_add_library module)
+            else()
+                # Static libraries have no linker PDBs, thus the compiler PDBs are relevant
+                set_target_properties(${target} PROPERTIES
+                                    COMPILE_PDB_NAME "${target}-s${SFML_PDB_POSTFIX}"
+                                    COMPILE_PDB_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}/lib")
```
```suggestion
                                      COMPILE_PDB_NAME "${target}-s${SFML_PDB_POSTFIX}"
                                      COMPILE_PDB_OUTPUT_DIRECTORY "${PROJECT_BINARY_DIR}/lib")
```

ChrisThrasher commented on src/SFML/System/CMakeLists.txt:
```diff
@@ -67,6 +67,12 @@ endif()
+# enable precompiled headers
+if (SFML_ENABLE_PCH)
+    message(VERBOSE "enabling PCH for SFML library 'sfml-system' (reused as the PCH for other SFML libraries)")
+    target_precompile_headers(sfml-system PRIVATE "${CMAKE_SOURCE_DIR}/src/SFML/PCH.hpp")
```
```suggestion
    target_precompile_headers(sfml-system PRIVATE ${PROJECT_SOURCE_DIR}/src/SFML/PCH.hpp)
```
`PROJECT_SOURCE_DIR` ensures that it's always starting from the root of SFML and not the root of whatever project may be including SFML via `add_subdirectory`.

ChrisThrasher reviewed (changes requested):
Can you share some stats for how this changes total CI runtime? Can we expect the pipeline to complete any quicker than before?

vittorioromeo commented:
@ChrisThrasher 

> Can you share some stats for how this changes total CI runtime? Can we expect the pipeline to complete any quicker than before?

I do not have any CI stats at the moment -- the goal of this PR is to add the PCH mechanism to SFML so that we can see how much of an improvement it is. I do expect a significant speedup for all Windows and Linux static library CI builds.

I have addressed your suggestions (thanks!) -- my suggestion would be to merge this PR, create a PR to enable PCH builds where appropriate in CI, and get some rough measurements. If there's no benefit, I'd be happy to revert this PR.

vittorioromeo commented:
SFMLBuildMaster: Build this please.

eXpl0it3r commented:
Thanks for this! 🙂  
As for the next steps, I'd suggest to select some builds. If we just do all or nothing, we're blind sighting ourselves in either direction. One could also consider enabling it on say GitHub Actions and leave it as is for Buildbot.

JonnyPtn commented:
What's the difference in performance if you don't reuse the PCH? I'm fairly sure that's the cause of the issues. So if we avoid doing that we can cover all platforms and build configurations?

Also solves the problem of adding more configuration - the typical way I've done this is to enable PCH by default then users can disable using `CMAKE_DISABLE_PRECOMPILE_HEADERS` without us having to do anything extra

vittorioromeo commented:
@JonnyPtn 

All the following results are on Windows 10 x64, clean static library build, build both tests and examples as well, timed via `time ninja`.

```bash
# no PCH at all
ninja  0.01s user 0.00s system 0% cpu 34.032 total # baseline

# PCH with reuse (this PR)
ninja  0.01s user 0.03s system 0% cpu 19.381 total # 15s improvement!

# PCH without reuse  
ninja  0.01s user 0.00s system 0% cpu 28.787 total # 6s improvement :/
```

So as you can see, reusing is quite significant, but without reusing we do still get a minor improvement. Maybe we could automatically pick reuse or no-reuse depending on whether we are building static or shared?

Also, I had to hack in `include_directories(${PROJECT_SOURCE_DIR}/src)` to make the `PCH.hpp` file find `WindowsHeader.hpp`. Not sure how to make that work cleanly, because that header is under `System`'s "private" headers.
